### ISSUE TITLE
How can I implement PPO with distributed sampling workers?

### ISSUE THREAD

**Issue by Bender1019** [seolhokim/DistributedRL-Pytorch-Ray#6]
@seolhokim  This repo is very helpful for me. My question is since there is impala which is central optimizing with data collected to the central buffer, and dppo which sync the communications between learner and actor. How can I run ppo with distributed sampling workers just like the partial combination between impala and dppo.  Thanks

**Comment by seolhokim** [comment]
@zyp57783
Thanks for watching my repository.
You understood the difference between the two algorithms well(maybe more than me). Learner(s) of Impala uses data(made by actors) that came from central buffer to update its network, and actors and learner(s) of DPPO communicate(send gradients or parameters) with each other through a parameter server.
But, I didn't understand "ppo with **distributed sampling workers**" Can you explain more about this? I guess you want to say use gradient and data together for update learner. Am I right?

**Comment by Bender1019** [comment]
Thanks for your reply. 
My point is that unlike DPPO of which the gradients are computed on each actors node, and  converged together through a parameter server, PPO with distributed sampling worker here I mean gradients are computed on the learner node with data coming from the central buffer.
Just like what baseline PPO does, except that data collection process become distributed among ray cluster.
I didn't mean I want to use both gradient and data to update learner, data is enough.
Last days I am trying to make this algorithm works based on your code. It's hard, I am not familiar
with Ray.

**Comment by seolhokim** [comment]
@zyp57783 
Good! you mean actors only collect data and save data to central buffer, and Learner(s) only gathers data from buffer and updates parameters, right? Then you need to think about three main problems. First, how to manage central buffer(shared resource management). you need to care about engineering things that I did like "What Replay buffer structure will be benefit to store data generated by actors and to use data by learner(s)?". Second, how to calculate reward(data will be collected from independent actors.), lastly, distribution between old policy and new policy(during training, old policy will gather data).

You can get another key point if you read more distributed RL paper. I'm not confident about ray, too. You will be able to do better than me :)

**Comment by Bender1019** [comment]
Thanks for suggestions.  I' ll try to do it.
